### PR TITLE
[RISCV] Prefer strided load for interleave load with only one lane active

### PR DESCRIPTION
**Pull request by preames**
If only one of the elements is actually used, then we can legally use a strided load in place of the segment load.  Doing so reduces vector register pressure, so if both segment and strided are believed to be element/segment at a time, then prefer the strided load variant.

Note that I've seen the vectorizer emitting wide interleave loads to represent a strided load, so this does happen in practice.  It doesn't matter much for small LMUL*NF, but at large NF can start causing problems in register allocation.

Note that this patch only covers the fixed vector formation cases.  In theory, we should do the same patch for scalable, but we can currently only represent NF2 in scalable IR, and NF2 is assumed to be optimized to better than segment-at-a-time by default, so there's currently nothing to do.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -21595,6 +21597,26 @@ bool RISCVTargetLowering::lowerInterleavedLoad(
...
+    Value *Offset = ConstantInt::get(XLenTy, Indices[0] * ScalarSizeInBytes);
+    Value *BasePtr = Builder.CreatePtrAdd(LI->getPointerOperand(), Offset);
+    Value *Mask = Builder.getAllOnesMask(VTy->getElementCount());
+    Value *VL = Builder.getInt32(VTy->getNumElements());
+
+    CallInst *CI = Builder.CreateIntrinsic(
+        Intrinsic::experimental_vp_strided_load,
+        {VTy, BasePtr->getType(), Stride->getType()},
+        {BasePtr, Stride, Mask, VL});
+    CI->addParamAttr(0, Attribute::getWithAlignment(CI->getContext(), LI->getAlign()));
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:21615]: 80 columns?

**Review by topperc — APPROVED** [review]
LGTM

**Comment by preames** [comment]
A quick note on an alternative approach, mostly for my future self.

The key semantic distinction between the load+shuffle and the strided load is the knowledge that the entire span of byte (even those not accessed) is dereferenceable.  However, for fixed length types (only), we can encode this fact in the dereferenceable argument attribute.  Given that, we could potentially do this as an IR canonicalization (i.e. always prefer strided for single use deinterleave), and then a backend could select the segment form for the strided instruction if profitable.  

This probably isn't worthwhile at the moment, but if we hit more cases where canonicalizing towards the strided form is useful - maybe strides greater than max NF? - we could explore this.  

This alternate approach doesn't work for scalable types.  Unless we add some kind of separate assume dereferenceable construct, or extend the attribute semantics to allow vscale constants.  Neither idea is crazy.

**Comment by lukel97** [comment]
> Note that I've seen the vectorizer emitting wide interleave loads to represent a strided load

I think long term we should teach the loop vectorizer it to emit strided loads when it has interleave groups with one member (in loop vectorizer parlance). That way we can properly cost model it too. 

@nikolaypanchenko was mentioning this a while ago, is there still plans to upstream this? https://github.com/llvm/llvm-project/pull/93972#issuecomment-2142655113

Mainly mentioning this because SLP can emit strided loads directly and it seems like that's the direction we should move in now that we have the VP strided load/store intrinsics

**Pull request merged**